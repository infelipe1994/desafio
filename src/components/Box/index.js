import styled from 'styled-components'
import { border, color, flexbox, layout, space } from 'styled-system'

export const Box = styled.div`
  ${border};
  ${color};
  ${flexbox};
  ${layout};
  ${space};
`
