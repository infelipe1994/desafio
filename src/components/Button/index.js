import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const Component = styled.button(
  ({ isFullWidth, isOutlined, theme }) => `
    appearance: none;
    align-items: center;
    background-color: ${isOutlined ? 'transparent' : theme.colors.green};
    border: 1px solid ${theme.colors.green};
    border-radius: 100rem;
    box-shadow: none;
    color: ${isOutlined ? theme.colors.green : theme.colors.smoke};
    cursor: pointer;
    display: inline-flex;
    font-size: ${theme.fontSizes.x1};
    font-weight: ${theme.fontWeights.x3};
    height: 2.5rem;
    justify-content: center;
    outline: none;
    padding: 0 ${theme.space.x4};
    position: relative;
    text-align: center;
    text-transform: uppercase;
    transition: all 0.250s;
    user-select: none;
    vertical-align: top;
    white-space: nowrap;

    &:hover, &:focus {
      background-color: ${theme.colors.green};
      border-color: ${theme.colors.green};
      color: ${theme.colors.smoke};
    }
    
    &:active {
      background-color: ${theme.colors.ocean};
      border-color: ${theme.colors.ocean};
      color: ${theme.colors.smoke};
    }

    &[disabled]{
      background-color: ${isOutlined ? 'transparent' : theme.colors.stone};
      border-color: ${theme.colors.stone};
      color: ${isOutlined ? theme.colors.stone : theme.colors.smoke};
      cursor: not-allowed;
    }

    ${isFullWidth ? 'display: flex; width: 100%' : ''};
  `
)

export const Button = p => <Component {...p} />

Button.propTypes = {
  isFullWidth: PropTypes.bool,
  isOutlined: PropTypes.bool
}
