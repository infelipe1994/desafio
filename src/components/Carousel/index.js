import React, { Children, useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { layout } from 'styled-system'

import { ControlButtonGroup } from '@/src/components/Carousel/ControlButtonGroup'
import { SlideGroup } from '@/src/components/Carousel/Slide/Group'

const Component = styled.div`
  ${layout};

  position: relative;
`

export const Carousel = ({
  children,
  currentSlide,
  onChangeSlide,
  ...rest
}) => {
  const childrenCount = Children.count(children) - 1
  const [slide, setSlide] = useState(currentSlide)
  const handleNext = () => {
    const nextSlide = slide === childrenCount ? childrenCount : slide + 1

    onChangeSlide(nextSlide)
    setSlide(nextSlide)
  }
  const handlePrevious = () => {
    const nextSlide = slide === 0 ? 0 : slide - 1

    onChangeSlide(nextSlide)
    setSlide(nextSlide)
  }

  useEffect(
    () => document.getElementById(`${currentSlide}`).scrollIntoView(),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  )

  return (
    <Component width={rest.width}>
      <SlideGroup {...rest}>{children}</SlideGroup>
      <ControlButtonGroup
        handleNext={handleNext}
        handlePrevious={handlePrevious}
        slide={slide}
      />
    </Component>
  )
}

Carousel.defaultProps = {
  currentSlide: 0,
  onChangeSlide: () => {},
  width: '100%'
}

Carousel.propTypes = {
  children: PropTypes.node,
  currentSlide: PropTypes.number,
  onChangeSlide: PropTypes.func,
  width: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    ),
    PropTypes.number,
    PropTypes.string
  ])
}
