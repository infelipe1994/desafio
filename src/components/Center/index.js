import PropTypes from 'prop-types'
import styled from 'styled-components'
import { layout, space } from 'styled-system'

export const Center = styled.div`
  ${layout};
  ${space};

  box-sizing: content-box;
  margin-left: auto;
  margin-right: auto;

  ${({ alsoCenterChildren }) =>
    alsoCenterChildren
      ? 'align-items: center; display: flex; flex-direction: column'
      : ''};

  ${({ alsoCenterText }) => (alsoCenterText ? 'text-align: center' : '')};
`

Center.defaultProps = {
  maxWidth: '40em'
}

Center.propTypes = {
  alsoCenterChildren: PropTypes.bool,
  alsoCenterText: PropTypes.bool,
  maxWidth: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    ),
    PropTypes.number,
    PropTypes.string
  ])
}
