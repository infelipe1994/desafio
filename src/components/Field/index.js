import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { theme } from '@/src/theme'
import { getColor } from '@/src/utilities/getColor'
import { isDateType } from '@/src/utilities/isDateType'
import { placeholder } from '@/src/utilities/placeholder'
import { uuid } from '@/src/utilities/uuid'

const Input = styled.input(
  ({ color }) => `
    appearance: none;
    background-color: transparent;
    border: 0;
    color: ${color};
    font-size: ${theme.fontSizes.x3};
    height: ${theme.lineHeights.x3};
    line-height: ${theme.lineHeights.x3};
    outline: 0;
    transition: all 0.250s;
    width: 100%;

    &[disabled] {
      cursor: not-allowed;
    }

    ${placeholder(color)};
  `
)

const Container = styled.div(
  ({ borderColor, error }) => `
    border-bottom: 1px ${error ? 'dashed' : 'solid'} ${borderColor};
    padding-bottom: ${theme.space.x1};
    position: relative;
    transition: all 0.250s;
    width: 100%;

    & input:not(:placeholder-shown) + label {
      opacity: 1;
      transform: translateY(-${theme.lineHeights.x3});
    }
  `
)

const Label = styled.label(
  ({ color, disabled }) => `
    bottom: ${theme.space.x1};
    color: ${color};
    font-size: ${theme.fontSizes.x3};
    left: 0;
    line-height: ${theme.lineHeights.x3};
    opacity: 0;
    position: absolute;
    transition: all 0.250s;

    ${disabled ? 'cursor: not-allowed' : ''};
  `
)

export const Field = ({ disabled, error, inputRef, label, type, ...rest }) => {
  const colors = getColor(disabled, error, theme)
  const id = useRef(rest.id || uuid())
  const isDate = isDateType(type)
  const stateProps = { ...colors, disabled, error }

  return (
    <Container {...stateProps}>
      <Input
        {...rest}
        {...stateProps}
        id={id.current}
        placeholder={label}
        ref={inputRef}
        type={isDate ? 'text' : type}
      />
      <Label {...stateProps} htmlFor={id.current}>
        {label}
      </Label>
    </Container>
  )
}

Field.defaultProps = {
  type: 'text'
}

Field.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  inputRef: PropTypes.func,
  label: PropTypes.string.isRequired,
  type: PropTypes.string
}
