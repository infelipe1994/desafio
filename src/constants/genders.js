export const genders = [
  { description: 'Feminino', value: 'female' },
  { description: 'Masculino', value: 'male' },
  { description: 'Outro', value: 'other' }
]
