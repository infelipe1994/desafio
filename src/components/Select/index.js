import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { Icon } from '@/src/components/Icon'
import { theme } from '@/src/theme'
import { getColor } from '@/src/utilities/getColor'
import { placeholder } from '@/src/utilities/placeholder'
import { uuid } from '@/src/utilities/uuid'

const Chevron = styled(Icon)`
  bottom: ${theme.space.x1};
  pointer-events: none;
  position: absolute;
  right: 0;
  transition: all 0.25s;
`

const Input = styled.select(
  ({ color }) => `
    appearance: none;
    background-color: transparent;
    border: 0;
    color: ${color};
    font-size: ${theme.fontSizes.x3};
    line-height: ${theme.lineHeights.x3};
    outline: 0;
    transition: all 0.250s;
    width: 100%;

    &[disabled] {
      cursor: not-allowed;
    }

    & option:checked {
      display: none;
    }

    ${placeholder(color)};
  `
)

const Container = styled.div(
  ({ borderColor, error, hasSelected }) => `
    border-bottom: 1px ${error ? 'dashed' : 'solid'} ${borderColor};
    display: flex;
    padding-bottom: ${theme.space.x1};
    position: relative;
    transition: all 0.250s;
    width: 100%;

    & ${hasSelected ? '' : 'select:active +'} label {
      opacity: 1;
      transform: translateY(-${theme.lineHeights.x3});
    }

    & select:active ~ svg {
      transform: rotate(180deg);
    }
  `
)

const Label = styled.label(
  ({ disabled, color, hasSelected }) => `
    bottom: ${theme.space.x1};
    color: ${color};
    font-size: ${theme.fontSizes.x3};
    left: 0;
    line-height: ${theme.lineHeights.x3};
    opacity: 1;
    position: absolute;
    transition: all 0.250s;

    ${disabled ? 'cursor: not-allowed' : ''};
    ${hasSelected ? '' : 'pointer-events: none'};
  `
)

export const Select = ({ disabled, error, inputRef, label, ...rest }) => {
  const colors = getColor(disabled, error)
  const hasSelected = !!(rest.defaultValue || rest.value)
  const id = useRef(rest.id || uuid())
  const stateProps = { ...colors, disabled, error, hasSelected }

  return (
    <Container {...stateProps}>
      <Input
        {...rest}
        {...stateProps}
        id={id.current}
        placeholder={label}
        ref={inputRef}
      />
      <Label {...stateProps} htmlFor={id.current}>
        {label}
      </Label>
      <Chevron color={colors.color} name="chevronDown" />
    </Container>
  )
}

Select.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  inputRef: PropTypes.func,
  label: PropTypes.string.isRequired
}
