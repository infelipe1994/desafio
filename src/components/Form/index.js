import React from 'react'
import PropTypes from 'prop-types'

import { Button } from '@/src/components/Button'
import { Stack } from '@/src/components/Stack'

export const Form = ({ children, disabled, onSubmit }) => (
  <Stack as="form" onSubmit={onSubmit} spaceBetween="x7">
    <Stack spaceBetween="x6">{children}</Stack>
    <Button disabled={disabled} isFullWidth>
      Avan√ßar
    </Button>
  </Stack>
)

Form.propTypes = {
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  onSubmit: PropTypes.func.isRequired
}
