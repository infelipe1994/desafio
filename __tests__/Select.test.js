import React from 'react'
import { cleanup, render } from '@testing-library/react'

import { Select } from 'src/components/Select'
import { genders } from 'src/constants/genders'

afterEach(cleanup)

describe('Select', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Select label="GÃªnero">
        {genders.map(gender => (
          <option key={gender.value} value={gender.value}>
            {gender.description}
          </option>
        ))}
      </Select>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
