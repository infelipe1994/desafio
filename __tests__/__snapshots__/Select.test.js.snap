// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Select__Container-sc-1fzd1i6-2 fHEPpu"
    color="#222222"
  >
    <select
      class="Select__Input-sc-1fzd1i6-1 kRRhPJ"
      color="#222222"
      id="1576b549-5774-4b57-876e-9f60b370edb6"
      placeholder="Gênero"
    >
      <option
        value="female"
      >
        Feminino
      </option>
      <option
        value="male"
      >
        Masculino
      </option>
      <option
        value="other"
      >
        Outro
      </option>
    </select>
    <label
      class="Select__Label-sc-1fzd1i6-3 fcdqBB"
      color="#222222"
      for="1576b549-5774-4b57-876e-9f60b370edb6"
    >
      Gênero
    </label>
    <svg
      aria-hidden="true"
      class="Select__Chevron-sc-1fzd1i6-0 hQZjlk"
      focusable="false"
      role="img"
      style="height: 1.000rem; width: 1.000rem;"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        fill="#222222"
      />
    </svg>
  </div>
</DocumentFragment>
`;
